src/kademlia/dht_state.cpp

```diff
@@ -0,0 +1,103 @@
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/kademlia/dht_state.hpp"
+
+#include <libtorrent/bdecode.hpp>
+#include <libtorrent/socket_io.hpp>
+
+namespace libtorrent {
+namespace dht
+{
+namespace
+{
+	node_id extract_node_id(bdecode_node const& e, string_view key)
+	{
+		if (e.type() != bdecode_node::dict_t) return node_id();
+		auto nid = e.dict_find_string_value(key);
+		if (nid.size() != 20) return node_id();
+		return node_id(nid);
+	}
+
+	void save_nodes(entry& ret, std::vector<udp::endpoint> const& v
+		, std::string const& key)
+	{
+		entry nodes(entry::list_t);
+		for (auto const& ep : v)
+		{
+			std::string node;
+			std::back_insert_iterator<std::string> out(node);
+			detail::write_endpoint(ep, out);
+			nodes.list().push_back(entry(node));
+		}
+		if (!nodes.list().empty())
+			ret[key] = nodes;
+	}
+} // anonymous namespace
+
+	dht_state read_dht_state(bdecode_node const& e)
+	{
+		dht_state ret;
+
+		if (e.type() != bdecode_node::dict_t) return ret;
+
+		ret.nid = extract_node_id(e, "node-id");
+#if TORRENT_USE_IPV6
+		ret.nid6 = extract_node_id(e, "node-id6");
+#endif
+
+		TORRENT_TRY {
+			if (bdecode_node const nodes = e.dict_find_list("nodes"))
+				detail::read_endpoint_list<udp::endpoint>(nodes, ret.nodes);
+		} TORRENT_CATCH(std::exception&) {}
+#if TORRENT_USE_IPV6
+		TORRENT_TRY{
+			if (bdecode_node const nodes = e.dict_find_list("nodes6"))
+				detail::read_endpoint_list<udp::endpoint>(nodes, ret.nodes6);
+		} TORRENT_CATCH(std::exception&) {}
+#endif
+
+		return ret;
+	}
+
+	entry save_dht_state(dht_state const& state)
+	{
+		entry ret(entry::dictionary_t);
+		ret["node-id"] = state.nid.to_string();
+		save_nodes(ret, state.nodes, "nodes");
```

Comment:
what about this idea, ``save_nodes`` returns an ``entry``. instead of passing in the name of the key you want it saved under, just assign it here: ``ret["nodes"] = save_nodes(state.nodes);``

would that work? (this of course assumes there is a move-assignment operator on ``entry``)